src/tactic/compute_degree.lean

```diff
@@ -0,0 +1,292 @@
+/-- `C_mul_terms e` produces a proof of `e.nat_degree = ??` in the case in which `e` is of the form
+`C a * X (^ n)?`.  It has special support for `C 1`, when there is a `nontrivial` assumption on the
+base-semiring. -/
+meta def C_mul_terms : expr → tactic unit
+| `(has_mul.mul %%C %%X) := match X with
+  | `(@polynomial.X %%R %%inst) := do
+      refine ``(polynomial.nat_degree_C_mul_X _ _),
+      assumption <|> exact ``(one_ne_zero)
+  | `(@has_pow.pow (@polynomial %%R %%nin) %%N %%inst %%mX %%n) := do
+      nontriviality_by_assumption R,
+      refine ``(polynomial.nat_degree_C_mul_X_pow %%n _ _),
+      assumption <|> exact ``(one_ne_zero)
+  | _ := trace "The leading term is not of the form\n`C a * X (^ n)`\n\n"
+  end
+| _ := fail "The leading term is not of the form\n`C a * X (^ n)`\n\n"
+
+--    nontriviality_by_assumption R,
+/--  Let `e` be an expression.  Assume that `e` is either a pure `X`-power or `C a` times a pure
+`X`-power in a polynomial ring over `R`.
+`single_term_resolve e` produces a proof of the goal `e.nat_degree = d`, where `d` is the
+exponent of `X`.
+
+Assumptions: either there is an assumption in context asserting that the constant in front of the
+power of `X` is non-zero, or the tactic `nontriviality R` succeeds. -/
+meta def single_term_resolve (e : expr) : tactic unit :=
+  match e.app_fn with
+  | `(coe_fn $ @polynomial.monomial %%R %%inst %%n) := do
+      refine ``(polynomial.nat_degree_monomial_eq %%n _),
+      assumption <|> exact ``(one_ne_zero)
+  | `(coe_fn $ @polynomial.C %%R %%inst) := do
+      exact ``(polynomial.nat_degree_C _)
+  | a := do C_mul_terms e <|>  -- either `e` is `C a * (X ^ n)` and `C_mul_terms e` handles it or
+    match e with
+    | `(@has_pow.pow (@polynomial %%R %%nin) %%N %%inst %%mX %%n) := do
+        nontriviality_by_assumption R,
+        refine ``(polynomial.nat_degree_X_pow %%n)
+    | `(@polynomial.X %%R %%inst) := do
+      nontriviality_by_assumption R,
+      exact ``(polynomial.nat_degree_X)
+    | _ := trace "The leading term is not of the form\n`C a * X (^ n)`\n\n"
+    --"`compute_deg` failed to compute the degree of the leading term\n\n"
+    end
+  end
+
+/--  Given an expression `e`, assuming it is a polyomial, `extract_deg_single_term e` tries to guess
+the `nat_degree` of `e`.  Currently, it supports:
+* `monomial n r`, guessing `n`,
+* `C a`, guessing `0`,
+* `polynomial.X`, guessing `1`,
+* `X ^ n`, guessing `n`,
+* everything else, guessing `e.nat_degree`.
+
+The expectation is that the argument of `extract_deg_single_term` is a factor of a summand of an
+expression in a polynomial ring. -/
+meta def extract_deg_single_term : expr → tactic expr
+| `(polynomial.X) := to_expr ``(1)
+| e := match e.app_fn with
+  | `(coe_fn $ polynomial.monomial %%n) := return n
+  | `(coe_fn $ polynomial.C) := to_expr ``(0)
+  | a := do e.get_app_args.nth 4 >>= return <|>
+         do val ← to_expr ``(polynomial.nat_degree),
+           return $ expr.mk_app val [e]
+  end
```

Comment:
Perhaps tidier as
```suggestion
meta def extract_deg_single_term : expr → tactic expr
| `(polynomial.X) := to_expr ``(1)
| `(polynomial.X ^ %%n) := return n
| (expr.app `(⇑(polynomial.monomial %%n)) x) := return n
| (expr.app `(⇑polynomial.C) x) := to_expr ``(0)
| e := to_expr ``(polynomial.nat_degree %%e)
```

Reply:
This is great!  Thank you so much!  I struggled a lot to get the matching to work and you saw the result!

Reply:
I think that, with this template, I can now simplify a few of the other matches in the file.